Sources/PCMultipartMessageEnricher.swift

```diff
@@ -0,0 +1,153 @@
+                sender: user,
+                room: self.room,
+                parts: basicMessage.parts,
+                createdAt: basicMessage.createdAt,
+                updatedAt: basicMessage.updatedAt
+            )
+            self.callCompletionHandlersForEnrichedMessagesWithIDsLessThanOrEqualTo(id: basicMessage.id, result: .success(message))
+        }
+    }
+    
+    fileprivate func callCompletionHandlersForEnrichedMessagesWithIDsLessThanOrEqualTo(id: Int, result: PCMultipartMessageEnrichmentResult) {
+        
+        // TODO: There may well be ways to make this faster
+        self.messageEnrichmentQueue.async(flags: .barrier) {
+            guard let nextIDToComplete = self.completionOrderList.first else {
+                self.logger.log("Message with id \(id) enriched but message enricher doesn't know about enriching it", logLevel: .debug)
+                return
+            }
+            
+            self.enrichedMessagesAwaitingCompletionCalls[id] = result
+            
+            guard id == nextIDToComplete else {
+                // If the message id received isn't the next to have its completionHandler called
+                // then return as we've already stored the result so it can be used later
+                self.logger.log(
+                    "Waiting to call completion handler for message id \(id) as there are other older messages still to be enriched",
+                    logLevel: .verbose
+                )
+                return
+            }
+            
+            repeat {
+                let messageID = self.completionOrderList.first!
+                
+                guard let completionHandler = self.messageIDToCompletionHandlers[messageID] else {
+                    self.logger.log("Completion handler not stored for message id \(messageID)", logLevel: .debug)
+                    return
+                }
+                
+                guard let result = self.enrichedMessagesAwaitingCompletionCalls[messageID] else {
+                    self.logger.log("Enrichment result not stored for message id \(messageID)", logLevel: .debug)
+                    return
+                }
+                
+                switch result {
+                case let .success(message):
+                    completionHandler(message, nil)
+                case let .error(err):
+                    completionHandler(nil, err)
+                }
+                
+                self.completionOrderList.removeFirst()
+                self.messageIDToCompletionHandlers.removeValue(forKey: messageID)
+                self.enrichedMessagesAwaitingCompletionCalls.removeValue(forKey: messageID)
+            } while self.completionOrderList.first != nil && self.enrichedMessagesAwaitingCompletionCalls[self.completionOrderList.first!] != nil
+        }
+    }
+}
+
+public enum PCMultipartMessageEnrichmentResult {
+    case success(PCMultipartMessage)
+    case error(Error)
+}
```

Comment:
There's a lot of duplication from `PCBasicMessageEnricher` here... and the enrichment process isn't conceptually any different since we don't care about the `parts` or the `text` in either case. Can we factor out the common stuff? Or even better make the enricher generic over both types? (Protocols?)

Reply:
Ahh I had this in the back of my mind as something to do at the end. Good you brought this up! Also at the same time I wasn't sure if I should start adding new things to the library or maintain the existing way of doing things..

Reply:
Although yeah, I think if I make this generic enough, then we can remove the other duplicated subscription you mentioned below..